In [1]:
import pandas as pd 

In [2]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

In [5]:
df = pd.read_csv(url)
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


### Check for Missing Values

We check each column for missing values before making any changes.

- `df.isna()` identifies missing values.
- `.sum()` counts the missing values in each column.

This helps us decide which columns need cleaning. This step does not change the data.

In [6]:
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Findings: Missing Values

Four columns contain missing values:

- `age`: 177 missing values.
- `embarked`: 2 missing values.
- `deck`: 688 missing values, the highest count.
- `embark_town`: 2 missing values.

All other columns have no missing values detected by `isna()`.
No data has been changed yet.

In [7]:
df.shape

(891, 15)

### Calculate the Percentage of Missing Values

The dataset has 891 rows and 15 columns.
We divide each column's missing count by the total number of rows
and multiply by 100 to calculate its missing percentage.

In [9]:
missing_percentage = (df.isna().sum() / len(df)) * 100 # i calculate ang percentage sa missing value
missing_percentage.round(2)

survived        0.00
pclass          0.00
sex             0.00
age            19.87
sibsp           0.00
parch           0.00
fare            0.00
embarked        0.22
class           0.00
who             0.00
adult_male      0.00
deck           77.22
embark_town     0.22
alive           0.00
alone           0.00
dtype: float64

### Finding: Missing Deck Information

The `deck` column has 688 missing values out of 891 records (77.22%).
Only 203 records have deck information.
We need to consider this limitation before using this column in analysis.

### Inspect Data Types

We use `df.info()` to inspect the columns, non-missing counts,
and data types. This helps us check whether each column uses
an appropriate data type before cleaning.

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    str    
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    str    
 8   class        891 non-null    str    
 9   who          891 non-null    str    
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    str    
 12  embark_town  889 non-null    str    
 13  alive        891 non-null    str    
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), str(7)
memory usage: 116.9 KB


### Findings: Data Types

The dataset contains integer, float, string, and boolean columns.
The `age` and `fare` columns are already numeric.
The `age` column has 714 non-missing values and 177 missing values.
No data types have been changed.

### Inspect Numeric Values

We use descriptive statistics to review numeric columns.
Minimum and maximum values help us identify values that may need investigation.

In [18]:
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Findings: Numeric Values

Recorded ages range from 0.42 to 80 years, with no negative ages.
The minimum age represents approximately 5 months.

Fares range from 0 to 512.3292.
Zero fares need further investigation before deciding whether they are errors.
No values have been changed.

In [36]:
df[df["fare"] == 0]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
179,0,3,male,36.0,0,0,0.0,S,Third,man,True,NaN,Southampton,no,True
263,0,1,male,40.0,0,0,0.0,S,First,man,True,B,Southampton,no,True
271,1,3,male,25.0,0,0,0.0,S,Third,man,True,NaN,Southampton,yes,True
277,0,2,male,NaN,0,0,0.0,S,Second,man,True,NaN,Southampton,no,True
302,0,3,male,19.0,0,0,0.0,S,Third,man,True,NaN,Southampton,no,True
413,0,2,male,NaN,0,0,0.0,S,Second,man,True,NaN,Southampton,no,True
466,0,2,male,NaN,0,0,0.0,S,Second,man,True,NaN,Southampton,no,True
481,0,2,male,NaN,0,0,0.0,S,Second,man,True,NaN,Southampton,no,True
597,0,3,male,49.0,0,0,0.0,S,Third,man,True,NaN,Southampton,no,True
633,0,1,male,NaN,0,0,0.0,S,First,man,True,NaN,Southampton,no,True


### Findings: Zero Fares

There are 15 records with a fare of zero across different ticket classes.
The available data does not explain why these fares are zero.
We retain these records because there is insufficient evidence to treat them as errors.

### Check for Possible Duplicate Rows

We count rows that repeat the same values across all columns.
Matching rows require review because different passengers may share the same recorded details.

In [43]:
df.duplicated().sum()

np.int64(107)

### Review Matching Rows

We found 107 repeated rows, excluding the first occurrence in each group.
We will inspect matching records before deciding whether to remove any.
The dataset has no passenger ID or name to confirm passenger identity.

In [47]:
df[df.duplicated(keep=False)].head(10)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
26,0,3,male,NaN,0,0,7.2250,C,Third,man,True,NaN,Cherbourg,no,True
28,1,3,female,NaN,0,0,7.8792,Q,Third,woman,False,NaN,Queenstown,yes,True
29,0,3,male,NaN,0,0,7.8958,S,Third,man,True,NaN,Southampton,no,True
32,1,3,female,NaN,0,0,7.7500,Q,Third,woman,False,NaN,Queenstown,yes,True
37,0,3,male,21.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
42,0,3,male,NaN,0,0,7.8958,C,Third,man,True,NaN,Cherbourg,no,True
45,0,3,male,NaN,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
46,0,3,male,NaN,1,0,15.5000,Q,Third,man,True,NaN,Queenstown,no,False
47,1,3,female,NaN,0,0,7.7500,Q,Third,woman,False,NaN,Queenstown,yes,True


In [48]:
df.loc[[32,47]]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
32,1,3,female,NaN,0,0,7.75,Q,Third,woman,False,NaN,Queenstown,yes,True
47,1,3,female,NaN,0,0,7.75,Q,Third,woman,False,NaN,Queenstown,yes,True


### Decision: Retain Matching Records

Rows 32 and 47 have identical recorded values, including missing age and deck.
Without passenger identifiers, we cannot confirm that matching rows represent the same person.
We retain all records to avoid removing potentially different passengers.

### Create a Working Copy

We create a separate copy for cleaning.
The original data remains in `df`, while cleaning changes will be made in `df_clean`.

In [49]:
df_clean = df.copy()
df.shape

(891, 15)

In [53]:
df_clean["deck"] = df_clean["deck"].fillna("Unknown")

df_clean["deck"].value_counts()

deck
Unknown    688
C           59
B           47
D           33
E           32
A           15
F           13
G            4
Name: count, dtype: int64

### Label Missing Embarkation Towns

We replace missing embarkation towns with "Unknown".
This keeps all records without guessing where these passengers boarded.

In [55]:
df_clean["embark_town"] = df_clean["embark_town"].fillna("Unknown")

df_clean["embark_town"].value_counts()

embark_town
Southampton    644
Cherbourg      168
Queenstown      77
Unknown          2
Name: count, dtype: int64

### Inspect Missing Embarkation Codes

We inspect records with missing `embarked` values and compare their
`embark_town` values before deciding how to handle them.

In [56]:
df_clean.loc[
    df_clean["embarked"].isna(),
    ["embarked", "embark_town"]

    ]

,embarked,embark_town
61,NaN,Unknown
829,NaN,Unknown


### Label Missing Embarkation Codes

Rows 61 and 829 have missing embarkation codes and unknown embarkation towns.
We label the missing codes as "Unknown" rather than guessing a boarding location.

In [60]:
df_clean["embarked"] = df_clean["embarked"].fillna("Unknown")

df_clean.loc[[61,829], ["embarked","embark_town"]]

,embarked,embark_town
61,Unknown,Unknown
829,Unknown,Unknown


### Preserve and Flag Missing Ages

We keep missing ages as NaN to avoid inventing passenger ages.
We add an `missing_age` column: True means the age is missing,
and False means an age is recorded.

In [63]:
df_clean["missing_age"] = df_clean["age"].isna()

df_clean["missing_age"].value_counts()

missing_age
False    714
True     177
Name: count, dtype: int64

### Validate Missing Values and Dataset Size

We check the remaining missing values and confirm that all 891 records remain.
Only `age` should contain missing values.
The added `missing_age` column identifies these records.

In [66]:
print("Dataset Shape:", df_clean.shape)
df_clean.isna().sum()

Dataset Shape: (891, 16)


survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
deck             0
embark_town      0
alive            0
alone            0
missing_age      0
dtype: int64

### Export the Cleaned Dataset

We save 891 records and 16 columns to `titanic_cleaned.csv`.
Unknown deck and embarkation information is labeled "Unknown".
Missing ages remain blank and are identified by `missing_age`.

In [67]:
df_clean.to_csv("titanic_cleaned.csv", index=False)

saved_df = pd.read_csv("titanic_cleaned.csv")
saved_df.shape

(891, 16)

### Cleaning Summary

- Preserved all 891 passenger records.
- Labeled 688 missing deck values as "Unknown".
- Labeled 2 missing values each in embarked and embark_town as "Unknown".
- Kept 177 missing ages and added the missing_age indicator.
- Retained zero fares and matching rows because they were not confirmed errors.
- Exported the dataset with 16 columns to titanic_cleaned.csv.

### Limitations

Unknown information remains unknown after labeling.
Age-based analysis uses only recorded ages unless an explicit
imputation method is applied. Matching records cannot be confirmed
as duplicate passengers without identifiers.